pageserver/src/tenant_mgr.rs

```diff
@@ -252,21 +250,73 @@ fn create_repo(
     wal_redo_manager: Arc<dyn WalRedoManager + Send + Sync>,
     remote_index: RemoteIndex,
 ) -> anyhow::Result<Arc<Repository>> {
-    let repo_dir = conf.tenant_path(&tenant_id);
-    ensure!(
-        !repo_dir.exists(),
-        "cannot create new tenant repo: '{}' directory already exists",
-        tenant_id
+    let target_tenant_directory = conf.tenant_path(&tenant_id);
+    anyhow::ensure!(
+        !target_tenant_directory.exists(),
+        "cannot create new tenant repo: '{tenant_id}' directory already exists",
+    );
+
+    let temporary_tenant_dir =
+        path_with_suffix_extension(&target_tenant_directory, TEMP_FILE_SUFFIX);
+    debug!(
+        "Creating temporary directory structure in {}",
+        temporary_tenant_dir.display()
     );
 
+    let temporary_tenant_timelines_dir = rebase_directory(
+        &conf.timelines_path(&tenant_id),
+        &target_tenant_directory,
+        &temporary_tenant_dir,
+    )?;
+    let temporary_tenant_config_path = rebase_directory(
+        &TenantConf::path(conf, tenant_id),
+        &target_tenant_directory,
+        &temporary_tenant_dir,
+    )?;
```

Comment:
I find all this rebasing complicated. I think this would actually be more readable if we just hardcode the directory structure here, and do something like:

```
let target_tenant_dir = format!("tenants/{tenant_id}");
let temp_tenant_dir = format!("tenants/{tenant_id}.temp");
let temp_tenant_config_path = format!("tenants/{tenant_id}.temp/config");
```

This isn't new with this PR I guess, many other places where we construct paths would be simpler if we just did the above, instead of using the various functions. But this is particularly verbose.

Reply:
Yes, I was also thinking in similar direction, but angling it as "specific rust API for our "file system"".
Might be good to present it as some small RFC, considering all FS operations we do and their properties.

Reply:
We can have some TenantDir structure or TenantPath with helper methods. TenantPath can be a newtype around PathBuf representing base dir, so you can set it up with arbitrary directory, whether it is a temp dir or a regular one

Hardcoding should happen in one place and currently this place is PageServerConf. Any solution that  avoids duplication of hardcoded values works for me